File: 0078-subsets/0078-subsets.py
class Solution:
    def subsetsHelper(self, nums, i, currSet, res):
        n = len(nums)
        
        if i >= n:
            res.append(currSet[:])
            return
        
        self.subsetsHelper(nums, i+1, currSet, res)
        
        currSet.append(nums[i])
        self.subsetsHelper(nums, i+1, currSet, res)
        currSet.pop()
            
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        currSet = []
        self.subsetsHelper(nums, 0, currSet, res)
        return res

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void subsetsHelper(vector<int> &nums, int i, vector<int> &curr, vector<vector<int>> & res){
        int n = nums.size();
        
        // base cases
        if(i>=n){
            res.push_back(curr);
            return;
        }
        
        //decisions 
        
        //1. Excluding 
        subsetsHelper(nums, i+1, curr, res);
        
        //2. Including 
        curr.push_back(nums[i]);
        subsetsHelper(nums, i+1, curr, res);
        curr.pop_back();
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        
        vector<int> curr;
        vector<vector<int>>res;
        
        subsetsHelper(nums, 0, curr, res);
        return res;
    }
};